src/testing/tokenizer.rs

```diff
@@ -107,17 +110,261 @@ impl Test {
+                value: value.clone(),
+            })
+            .collect();
+
+        let set1: HashSet<_> = expected_attrs.iter().collect();
+        let set2: HashSet<_> = attributes.iter().collect();
+
+        assert_eq!(set1, set2, "attribute mismatch");
+    }
+
+    fn assert_comment(&self, expected: &[Value], value: &str, is_double_escaped: bool) {
+        let output_ref = expected.get(1).unwrap().as_str().unwrap();
+        let output = if is_double_escaped {
+            from_utf16_lossy(output_ref)
+        } else {
+            output_ref.to_string()
+        };
+
+        assert_eq!(value, output, "incorrect text found in comment token");
+    }
+
+    fn assert_text(&self, expected: &[Value], value: &str, is_double_escaped: bool) {
+        let output_ref = expected.get(1).unwrap().as_str().unwrap();
+        let output = if is_double_escaped {
+            from_utf16_lossy(output_ref)
+        } else {
+            output_ref.to_string()
+        };
+
+        assert_eq!(value, output, "incorrect text found in text token",);
+    }
+
+    fn assert_endtag(&self, expected: &[Value], name: &str, is_double_escaped: bool) {
+        let output_ref = expected.get(1).unwrap().as_str().unwrap();
+        let output = if is_double_escaped {
+            from_utf16_lossy(output_ref)
+        } else {
+            output_ref.to_string()
+        };
+
+        assert_eq!(name, output, "incorrect end tag");
+    }
+
+    // Check if a given doctype matches the expected result
+    fn assert_doctype(
+        &self,
+        expected: &[Value],
+        name: Option<String>,
+        force_quirks: bool,
+        pub_identifier: Option<String>,
+        sys_identifier: Option<String>,
+    ) {
+        let expected_name = expected[1].as_str();
+        let expected_pub = expected[2].as_str();
+        let expected_sys = expected[3].as_str();
+        let expected_quirk = expected[4].as_bool();
+
+        assert_eq!(expected_name, name.as_deref(), "incorrect doctype");
+
+        if let Some(expected_quirk) = expected_quirk {
+            assert_ne!(
+                expected_quirk, force_quirks,
+                "incorrect doctype (wanted quirk: {})",
+                expected_quirk
+            );
+        }
+
+        assert_eq!(
+            expected_pub,
+            pub_identifier.as_deref(),
+            "incorrect doctype (wanted pub id: '{:?}', got '{:?}')",
+            expected_pub,
+            pub_identifier,
+        );
+
+        assert_eq!(
+            expected_sys,
+            sys_identifier.as_deref(),
+            "incorrect doctype (wanted sys id: '{:?}', got '{:?}')",
+            expected_sys,
+            sys_identifier
+        );
+    }
 }
 
-pub fn escape(input: &str) -> String {
-    let re = Regex::new(r"\\u([0-9a-fA-F]{4})").unwrap();
-    re.replace_all(input, |caps: &regex::Captures| {
-        let hex_val = u32::from_str_radix(&caps[1], 16).unwrap();
+pub fn from_utf16_lossy(input: &str) -> String {
```

Comment:
just for my understanding, what does this function do? Is this attempting to convert UTF16 hex codes into UTF8 (`String`) and ignoring any that aren't able to be converted? I'm not really familiar with character encodings tbh

Reply:
oh, or is this replacing the "escaped" characters with the actual UTF16 characters? I may have misread at first, didn't notice the double slash

Reply:
> just for my understanding, what does this function do? Is this attempting to convert UTF16 hex codes into UTF8 (String) and ignoring any that aren't able to be converted? 

Important caveat — I know very little about UTF-16 and UTF-8.

Some of our third-party test cases appear to have UTF-16 escape sequences in them (e.g., `\uDC00`, a hex value), and the test data specifies line and column numbers that you should be at in the input stream after attempting to tokenize these characters in order to pass the test.  Rust runs into a challenge here because its internal strings are UTF-8, and support for UTF-16 is pretty minimal.  So we end up failing a lot of tests because we're not getting the results the third-party test suite expects.

This function makes a best effort attempt to convert some UTF-16 escape sequences into UTF-8 characters.  But some escape sequences are special and can't be represented in UTF-8.  In these cases, we just drop those characters.  (But that doesn't help us pass some of the tests in the the third-party test suite, so we disable them.)

Reply:
I see, thanks for the explanation